flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/wmassigners/WatermarkAssignerOperatorFactory.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.operators.wmassigners;
+
+import org.apache.flink.streaming.api.graph.StreamConfig;
+import org.apache.flink.streaming.api.operators.ChainingStrategy;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperatorFactory;
+import org.apache.flink.streaming.api.operators.Output;
+import org.apache.flink.streaming.api.operators.StreamOperator;
+import org.apache.flink.streaming.runtime.tasks.StreamTask;
+import org.apache.flink.table.dataformat.BaseRow;
+import org.apache.flink.table.runtime.generated.GeneratedWatermarkGenerator;
+
+/**
+ * The factory of {@link WatermarkAssignerOperator}.
+ */
+public class WatermarkAssignerOperatorFactory implements OneInputStreamOperatorFactory<BaseRow, BaseRow> {
+	private static final long serialVersionUID = 1L;
+
+	private final int rowtimeFieldIndex;
+
+	private final long idleTimeout;
+
+	private final GeneratedWatermarkGenerator generatedWatermarkGenerator;
+
+	private ChainingStrategy strategy = ChainingStrategy.HEAD;
+
+	public WatermarkAssignerOperatorFactory(
+			int rowtimeFieldIndex,
+			long idleTimeout,
+			GeneratedWatermarkGenerator generatedWatermarkGenerator) {
+		this.rowtimeFieldIndex = rowtimeFieldIndex;
+		this.idleTimeout = idleTimeout;
+		this.generatedWatermarkGenerator = generatedWatermarkGenerator;
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public StreamOperator createStreamOperator(StreamTask containingTask, StreamConfig config, Output output) {
+		WatermarkGenerator watermarkGenerator = generatedWatermarkGenerator.newInstance(containingTask.getUserCodeClassLoader());
```

Comment:
You can code generate this during `WatermarkAssignerOperator::open`? So we can avoid introducing this factory

Reply:
Yes, we can. But I think the factory way is more recommended by the community. This can make the `watermarkGenerator` to be final in the member area and make the logic of `WatermarkAssignerOperator` clean (not cooperate with so many variables). 

`AsyncWaitOperator` is also refactored to `AsyncWaitOperatorFactory`.

Reply:
I'm not sure whether factory is more recommended way since I don't see any deprecation comment in transform with operator. 
My main concern about this is other code generated operators are all generating classes in `open` method, thought would be good to keep align with others.

Reply:
I think we should gradually refactor existing generated operators. Take `AggregateWindowOperator` as an exmple, every generated class has two objects, e.g. `aggWindowAggregator` and `generatedAggWindowAggregator`, etc.. And they are all nullable, and we have to add a lot of `if else` to compile them in `open` method. 

I think if we have a `WindowOperatorFactory`, the logic and member fields will be simpler.